free/src/main/scala/cats/free/FreeT.scala

```diff
@@ -0,0 +1,254 @@
+package cats
+package free
+
+import scala.annotation.tailrec
+
+import cats.data.Xor
+
+/** FreeT is a monad transformer for Free monads over a Functor S
+  */
+sealed abstract class FreeT[S[_], M[_], A] extends Product with Serializable {
+
+  import FreeT._
+
+  final def map[B](f: A => B)(implicit M: Applicative[M]): FreeT[S, M, B] =
+    flatMap(a => pure(f(a)))
+
+  /** Binds the given continuation to the result of this computation. */
+  final def flatMap[B](f: A => FreeT[S, M, B]): FreeT[S, M, B] =
+    Gosub(this, f)
+
+  /**
+   * Changes the underlying `Monad` for this `FreeT`, ie.
+   * turning this `FreeT[S, M, A]` into a `FreeT[S, N, A]`.
+   */
+  def hoist[N[_]](mn: M ~> N): FreeT[S, N, A] =
+    step match {
+      case e @ Gosub(_, _) =>
+        Gosub(e.a.hoist(mn), e.f.andThen(_.hoist(mn)))
+      case Suspend(m) =>
+        Suspend(mn(m))
+    }
+
+  /** Change the base functor `S` for a `FreeT` action. */
+  def interpret[T[_]](st: S ~> T)(implicit M: Functor[M]): FreeT[T, M, A] =
+    step match {
+      case e @ Gosub(_, _) =>
+        Gosub(e.a.interpret(st), e.f.andThen(_.interpret(st)))
+      case Suspend(m) =>
+        Suspend(M.map(m)(_.map(s => st(s))))
+    }
+
+  /**
+    * Runs to completion, mapping the suspension with the given transformation
+    * at each step and accumulating into the monad `M`.
+    */
+  def foldMap(f: S ~> M)(implicit MR: MonadRec[M]): M[A] = {
+    def go(ft: FreeT[S, M, A]): M[FreeT[S, M, A] Xor A] =
+      ft match {
+        case Suspend(ma) => MR.flatMap(ma) {
+          case Xor.Left(a) => MR.pure(Xor.Right(a))
+          case Xor.Right(sa) => MR.map(f(sa))(Xor.right)
+        }
+        case g @ Gosub(_, _) => g.a match {
```

Comment:
The abstract type `A` in `GoSub` in which both `a` and `f` depend seem to cause inference to not work properly unless the whole thing is named :P. 
This technique is also used in scalaz. Open to changes if anyone thinks there is a better way. 
https://github.com/scalaz/scalaz/pull/730